x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/pivot/Pivot.java

```diff
@@ -189,44 +189,37 @@ public boolean supportsIncrementalBucketUpdate() {
         );
     }
 
-    public QueryBuilder filterBuckets(Map<String, Set<String>> changedBuckets) {
-
-        if (changedBuckets == null || changedBuckets.isEmpty()) {
-            return null;
-        }
+    public QueryBuilder filterBuckets(
+        Map<String, Set<String>> changedBuckets,
+        String synchronizationField,
+        long lastSynchronizationCheckpoint
+    ) {
 
         if (config.getGroupConfig().getGroups().size() == 1) {
             Entry<String, SingleGroupSource> entry = config.getGroupConfig().getGroups().entrySet().iterator().next();
-            // it should not be possible to get into this code path
-            assert (entry.getValue().supportsIncrementalBucketUpdate());
 
             logger.trace("filter by bucket: " + entry.getKey() + "/" + entry.getValue().getField());
-            if (changedBuckets.containsKey(entry.getKey())) {
-                return entry.getValue().getIncrementalBucketUpdateFilterQuery(changedBuckets.get(entry.getKey()));
-            } else {
-                // should never happen
-                throw new RuntimeException("Could not find bucket value for key " + entry.getKey());
-            }
+            Set<String> changedBucketsByGroup = changedBuckets.get(entry.getKey());
+
+            // important: the fields must match to apply this optimization
+            long synchronizationTimestamp = entry.getKey().equals(synchronizationField) ? lastSynchronizationCheckpoint : 0;
```

Comment:
It seems we are needlessly adding an if branch. I think we should probably just supply both the `synchronizationField` and the `lastSynchronizationCheckpoint` to `getIncrementalBucketUpdateFilterQuery`. 

(similar comment for the more than one group source entry)

Reply:
:+1: This is a bug, `entry.getKey()` returns the name of the group, not the field that is used (entry.getValue().getField()), but I think I will follow your suggestion and pass it in and let the GroupSource decide what to do.